### ISSUE TITLE
`sqrt` vs `num-sqrt`

### ISSUE THREAD

**Issue by shriram** [brownplt/pyret-lang#811]
```
> num-sqrt(-1)
sqrt: negative argument -1
```
Note the name change.

**Comment by blerner** [comment]
This one's a dumb bug.  The short answer is, I spent a fair bit of time plumbing better error callbacks through the js-numbers library, so that errors returned by different runtimes were properly different (this scenario arises in the compiler tests, where we exec newly-compiled code).  But, for compatibility, I didn't change the *implementation* of what those callbacks do: they still just call `throwMessageException`.  Several invocations of the callbacks include some text trying to claim which function threw the error, and in js-numbers, `sqrt` is named `sqrt`, so that's correct, from its perspective.

@jswrenn the remaining bits of this are error-variants and error-rendering improvements that should likely fall in your area of expertise.  Look at the definitions of `NumberErrbacks` in runtime.js, fix them to throw new error variants, and then clean up their usages in js-numbers to match whatever new information you throw.  (There's not much to work with, since you're deep in runtime functions that only have values, not srclocs or other information, as arguments.  I suspect most of these errbacks should be nullary functions, as a result.)

**Comment by sorawee** [comment]
There is a bigger problem in here:

1. spyret also uses this function with the name `sqrt`. If we change it to `num-sqrt`, spyret would get the wrong name (in fact, `sqrt` is not a good example, since spyret has complex number, so the functions are not really the same, but we could imagine one where the problem actually exists)
2. The problem doesn't apply to only jsnumbers. Take `within` as an example:

    ```
    > within(1)
    <function:within-rel(...)>
    > within-rel(1)
    <function:within-rel(...)>
    ```

Fixing `sqrt` might be easy, but the core problem remains.

See also #503

**Comment by blerner** [comment]
Not quite, that's the goal of the NumberErrbacks -- you can throw the appropriate error for the appropriate scenario.

The fact that within and within-rel are _deliberately_ aliased was a design choice we made; that's incidental to the problem being raised here.  Shriram's original point is mostly moot; the relevant problem is we're still throwing message exceptions, when we have the plumbing and the ability to throw richer errors.

**Comment by blerner** [comment]
See #758

**Comment by Emmay** [comment]
This code now gives the following, much more helpful error:

```
The function application
num-sqrt(-1)
failed because the 1ˢᵗ argument evaluated to an unexpected value.
An annotation, NumNonNegative, in <builtin Numbers>
was not satisfied by the value -1

```
Since #758 is still open, and this specific error seems to be fixed, is this closable?
